### PR TITLE
Audit Log Fix

### LINKED ISSUE #18: [Bug]: Automod actions get logged incorrectly when using audit logs
### Checked Existing

- [X] I have checked the repository for duplicate issues.

### What happened?

When the automod does actions like timing someone out, which the bot logs the incorrect user as having executed the action

### What did you expect to happen?

The bot to log that it was the automod. Time checks need to be added for all instances of audit log checking

### Steps to reproduce?

This happens because some checks the bot does rely on heuristics rather than data from Discord. For example when timing someone out, the generic "member updated" event is sent. The bot checks to see if `communicationDisabledUntilTimestamp` has been changed, and if so then it knows the update was for a timeout. To know who timed the user out, the bot will then check the audit log for the most recent `MEMBER_UPDATE` log, and use the executor of that in the log. However in this case a time check is not done on the audit log, and the automod does not create an audit log. So when an automod times a user out, the bot will pull the log for the most recent real user who timed someone out

### Other relevant information. (OPTIONAL)

_No response_

### PR DESCRIPTION
**Pull request by wolfendale**
Resolves #18, #12 

### Changes:

Previously we were creating event log entries by listening for specific events and then attempting to find an appropriate log in the audit log to provide extra information about the action (who banned someone, why, etc) This caused issues when an appropriate audit log entry could not be found for the change, causing some event logs to have the incorrect information associated with them.

As we've now upgraded to discord.js v14 we have access to the `GuildAuditLogEntryCreate` event, which means for most of the events that we want to log, we can create them directly from the audit logs in discord. This PR updates Chubby to use this technique to handle the following events:
- Member timeouts
- Member nickname changes
- Member kicks
- Member bans

Unfortunately there is an event log that we create when a member leaves of their own accord which no audit log is created for as this is not a moderation/administration action. This PR continues to use the current handling for this case. This does mean there's a potential race condition that would mean that when a user is kicked or banned, we create two event logs, one for the kick/ban and one generic "Member Left" event, however we shouldn't miss any events, only potentially see duplicates.

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [x] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).
- [x] I have tested all of my changes.

### CODE REVIEW

File: src/events/guildAuditLogEntryCreate.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+				if (timeoutChange.new && timeoutChange.new && typeof timeoutChange.new === 'string') {
+					const timeout = new Date(timeoutChange.new);
+					await handleMemberTimedOut(guild, user, executor, auditLogEntry.reason, timeout);
+				}
+			}
+
+			// * This checks that a user has changed their name
+			const nickChange = auditLogEntry.changes.find(change => change.key === 'nick');
+			// * The types here should always be correct, but I have no idea how to convince the discord.js types of that...
+			if (nickChange && typeof nickChange.new === 'string' && (nickChange.old === undefined || typeof nickChange.old === 'string')) {
```
Comment (jonbarrow) [inline src/events/guildAuditLogEntryCreate.ts:33]: Why the check for both `undefined` or `string`? That feels redundant, since it would always be one of those. It reads like checking `if (true || false)` to me

Also what isses with types were you having? TypeScript seems perfectly happy to narrow these types in my test:

Screenshot from 2024-07-19 08-30-43

File: src/events/guildAuditLogEntryCreate.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+	await sendEventLogMessage(guild, null, embed);
+}
+
+async function handleMemberKick(auditLogEntry: GuildAuditLogsEntry, guild: Guild): Promise<void> {
+	// * Unsure why just checking the action type doesn't narrow the type here, probably too much type inference going on?
+	const user = (auditLogEntry as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberKick>).target;
+	if (!user) {
+		return;
+	}
```
Comment (jonbarrow) [inline src/events/guildAuditLogEntryCreate.ts:133]: Explicitly typing the entry does narrow the type:

Screenshot from 2024-07-19 08-33-37

File: src/events/guildMemberRemove.ts [hunk 3]
```diff
@@ -13,16 +13,28 @@ export default async function guildMemberRemoveHandler(member: GuildMember | Par
 	const user = member.user;
 	
 	const auditLogs = await guild.fetchAuditLogs({
-		limit: 1
+		limit: 10
```
Comment (jonbarrow) [inline src/events/guildMemberRemove.ts:16]: Why the increase?

File: src/events/guildMemberRemove.ts [hunk 4]
```diff
@@ -13,16 +13,28 @@ export default async function guildMemberRemoveHandler(member: GuildMember | Par
...
 	});
 
+	const auditLog = auditLogs.entries.find(log => {
+		return (log.action === AuditLogEvent.MemberBanAdd || log.action === AuditLogEvent.MemberKick) &&
+		log.targetId === member.id &&
+		Date.now() - log.createdTimestamp < 2000;
+	});
```
Comment (jonbarrow) [inline src/events/guildMemberRemove.ts:23]: Not an "issue" but just from a stylistic standpoint I try to avoid stuff like this (both the very long `if` style checks with many conditions, and directly returning on them). It makes it a bit harder to parse at a glance imo

This can be refactored to use a fix of `filter` to narrow the log actions, and then `find` to get the correct one

File: src/events/guildMemberRemove.ts (hunk 3 above)
Comment (wolfendale) [reply src/events/guildMemberRemove.ts:16]: We can't limit the number of audit logs, we just filter them afterwards, if there was another most recent audit log here we'd miss it, so we may as well bring back some more. I don't have any evidence to show what a better number for this is

File: src/events/guildMemberRemove.ts (hunk 4 above)
Comment (wolfendale) [reply src/events/guildMemberRemove.ts:23]: The filter is still going to be long, do you want to break the `&&`d parts up into separate filters, or would you still `&&` those together on the `filter`?
```ts
auditLogs.entries
    .filter(log => Date.now() - log.createdTimestamp < 2000)
    .filter(log => log.action === AuditLogEvent.MemberBanAdd || log.action === AuditLogEvent.MemberKick)
    .find(log => log.targetId === member.id);
```

File: src/events/guildAuditLogEntryCreate.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
+import { AuditLogEvent, EmbedBuilder } from 'discord.js';
+import { sendEventLogMessage } from '@/util';
+import type { Guild, User, GuildAuditLogsEntry } from 'discord.js';
+
+export default async function guildAuditLogEntryCreateHandler(auditLogEntry: GuildAuditLogsEntry, guild: Guild): Promise<void> {
+	switch (auditLogEntry.action) { 
+		case AuditLogEvent.MemberUpdate: {
+
+			// * Unsure why just checking the action type doesn't narrow the type here, probably too much type inference going on?
+			const user = (auditLogEntry as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>).target;
```
[the remark this answers is not in the document]
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts:10]: So I tried this, unfortunately the types are _really messed up_.

In theory, by asserting that `auditLogEntry.action === ActionLogEvent.MemberUpdate` (which we're doing in the switch) then the type of `auditLogEntry` should be inferred as `GuildAuditLogsEntry<ActionLogEvent.MemberUpdate>`. Unfortunately the type of '`auditLogEntry` doesn't get refined by the `switch` it doesn't even get refined if I change this into a chained `if`.

I tried creating a custom type predicate and that wasn't working, but I've just tried again and it works, so fuck me I guess. I'll update

File: src/events/guildAuditLogEntryCreate.ts (hunk 1 above)
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts:33]: `old` and `new` for a nickname change _should_ have the type `string | undefined` because of the nature of the change. But the type doesn't get narrowed even though we know what kind of change event it is, and since `AuditLogChange` isn't parameterised on a type, we can't specialise it any more.

That means without a cast or a guard, `old` and `new` have a type like this:
image

You can even do things like:
```ts
		const x: AuditLogChange = {
			key: 'nick',
			old: 1,
			new: 2
		};
```
And that's _totally_ valid. Even though the types should be able to be narrowed. This is because `AuditLogChange` looks like this:

```ts
export interface AuditLogChange {
  key: APIAuditLogChange['key'];
  old?: APIAuditLogChange['old_value'];
  new?: APIAuditLogChange['new_value'];
}
```
Where `APIAuditLogChange` is the union type of all the different interfaces that an audit change can be.

I'm not even sure how we'd write a type guard for this because there's not even a value to compare, only types.

File: src/events/guildAuditLogEntryCreate.ts (hunk 2 above)
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts:133]: Solved with type guard from above

File: src/events/guildAuditLogEntryCreate.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: Can you post the actual TypeScript errors your are getting here? I'm trying to replicate the issue but I can't my tests are narrow the type down correctly:

Screenshot from 2024-07-19 19-12-34

File: src/events/guildAuditLogEntryCreate.ts (hunk 1 above)
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: Yeah I agree, Discord.JS's types are kinda fucked. To the point where I honestly sort of want to roll our own minimal client just for us with better types, because it's getting ridiculous. I've taken a look too now

Anyway, this check can still be cleaned up a bit by only checking the `typeof === 'string'`. If it's `undefined` then that will be false, which ensures that we have the correct values and that they exist:

```ts
import { type GuildAuditLogsEntry, AuditLogEvent } from 'discord.js';

function test(auditLogEntry: GuildAuditLogsEntry) {
	switch (auditLogEntry.action) {
		case AuditLogEvent.MemberUpdate:
			const nickChange = auditLogEntry.changes.find(change => change.key === 'nick');

			if (nickChange && typeof nickChange.old === 'string' && typeof nickChange.new === 'string') {
				test2(nickChange.old);
				test2(nickChange.new);
			}
			break;
	}
}

// Simulate a function that MUST have a string, not undefined/null
function test2(str: string) {

}
```

File: src/events/guildAuditLogEntryCreate.ts (hunk 2 above)
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: I'm not sure what you mean? This doesn't look changed, and by explicitly typing `auditLogEntry` here you can remove the `as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberKick>`, since it will already be narrowed, like I showed in my screenshot?

File: src/events/guildMemberRemove.ts (hunk 3 above)
Comment (jonbarrow) [reply src/events/guildMemberRemove.ts:16]: That's fair. It's very very unlikely, but it's true that it's a race condition

File: src/events/guildMemberRemove.ts (hunk 4 above)
Comment (jonbarrow) [reply src/events/guildMemberRemove.ts:23]: Using what you suggested seems fine, using two `filter`s and a `find` breaks it up into more readable components which clearly describe the checks at a glance

File: src/events/guildAuditLogEntryCreate.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts]: In your example here, the type of `user` isn't getting narrowed? It's still any of that set of union types. I want it to be a `User | null` which is what it would be in a `GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>`

I've changed from the switch to a series of `if` / `else`s using the type predicate to narrow the type so that the type of the event is narrowed, giving the right types for the target

image

Example of the error:
image

File: src/events/guildAuditLogEntryCreate.ts (hunk 1 above)
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts]: In theory both of these should be able to be undefined because you can remove your server nick / add one without having one before, so in theory we need the undefined check for both

File: src/events/guildAuditLogEntryCreate.ts (hunk 2 above)
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts]: I don't know if this is just GitHub being weird? I've removed all of the casts from every `user` instance in the file, and I didn't even rebase/force push

File: src/events/guildAuditLogEntryCreate.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: I was referring to the `auditLogEntry` type, not the `user` type. The type assertions you're doing are on `auditLogEntry` in order to get the `target`, but it already has access to the `target` field at this point

If what you actually want is to narrow the `user` type, you can simply do that on the `user` itself:

```ts
import {
	AuditLogEvent,
	type GuildAuditLogsEntry,
	type User
} from 'discord.js';

function test(auditLogEntry: GuildAuditLogsEntry) {
	switch (auditLogEntry.action) {
		case AuditLogEvent.MemberUpdate:
			const user = auditLogEntry.target as User | null;

			if (!user) {
				return;
			}

			handleUser(user);
			break;
	}
}

function handleUser(user: User) {
	// Do something
}
```

The concern I was pointing out was mostly centered around the use of `as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>`, which is hard to parse at a glance and (not your fault, it's the libs fault), honestly just ugly. Asserting on `user` is much more clear and shorter

File: src/events/guildAuditLogEntryCreate.ts (hunk 1 above)
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: Ah that's right, I forgot about that. In that case, this doesn't seem to allow for `new` to be undefined?

File: src/events/guildAuditLogEntryCreate.ts (hunk 2 above)
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: It must be an issue with the mobile app, on desktop I do see the changes. Odd

File: src/events/guildAuditLogEntryCreate.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts]: Are you ok with the way that I've done this currently? Using a type predicate to narrow the audit log type, then I don't need to cast `user` at all

e.g.
```ts
	if (logIsForEvent(auditLogEntry, AuditLogEvent.MemberUpdate)) {
		const user = auditLogEntry.target;
		if (!user) {
			return;
		}
                ...
```

File: src/events/guildAuditLogEntryCreate.ts (hunk 1 above)
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts]: It'd be pretty ugly and hard to read if I inlined this so I've added a type predicate to tidy things up:
```ts
if (nickChange && isStringOrUndefined(nickChange.old) && isStringOrUndefined(nickChange.new)) {
	await handleMemberNicknameChange(guild, user, nickChange.old, nickChange.new);
}

function isStringOrUndefined(input: any): input is string | undefined {
	return input === undefined || typeof input === 'string';
}
```

You ok with that?

Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: That's perfect :+1:

File: src/events/guildAuditLogEntryCreate.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonbarrow) [reply src/events/guildAuditLogEntryCreate.ts]: Yes that's fine :+1: you say "currently", was this already in place? I think that the GitHub mobile app may have been playing tricks on me last night when I did this review...

[the remark this answers is not in the document]
Comment (wolfendale) [reply src/events/guildAuditLogEntryCreate.ts]: Yeah, it was already like this 😂
